### ISSUE TITLE
[WIP] Draggable Dialog

### ISSUE THREAD

**Issue by AlexejLiebenthal** [MaterialDesignInXAML/MaterialDesignInXamlToolkit#1196]
As I know this is not part of Material Design spec.
But if you like to implement it in web, you have easy possibilities to do this with either [React](https://material-ui.com/demos/dialogs/#draggable-dialog) or [Angular](https://stackoverflow.com/a/49706221/3924697).

For phones I wouldn't like that feature, as you don't have any space to drag the dialog around, but for desktop applications or even on tablets this feature with a wider screens, makes sense, if you maybe like to see what was behind the dialog and it is somehow also Desktop-Application default/expected behavior for dialogs.

## Current implementations
There are two styles, which can be applied to the dialog host. 
1. (default) - this style is based on a `Popup` (`PopupEx`) for the dialog.
2. `MaterialDesignEmbeddedDialogHost` - this style is based on `Grid` for the dialog. 

## Current behavior
The current behavior is, ...
- that a `Card` is set as the `Grid` or `Popup` content.
- that the dialog will be shown centered to the dialog host.
- that the content (`DialogHost.DialogContent`, `DialogHost.DialogContentTemplate` etc.) is shown in the Card content.
- that while resizing/dragging the dialog remains centered.

### default
- [ ] As implemented as a popup the drag could be done anywhere on the screen and is not limited to the dialog host. Should this be allowed or should it be limited?
- [ ] Where should the dialog move on resize or drag of the host window? I think `PopupEx` has already the right behavior, if the dialog is drawn in the window. But if you drag the popup out of the window boundaries should there still be a repositioning?
### MaterialDesignEmbeddedDialogHost
- [ ] The drag here could be realized with margins. In opposite to the default style, this is already bounded to the dialog host content or maybe window content. 
- [ ] As working with margins, on resize the dialog could disappear. Here we need to develop a resize/maximize/restore strategy.
### both
- [ ] What should happen, if the dialog reopens? Should the last position persist (this would be kind of the default behavior, as the properties of the `Grid`/`Popup` would be modified on drag)?

**Comment by Keboo** [comment]
First of all thank you for this nice write up @AlexejLiebenthal 

A few thoughts I have:
- There seems to be a subtle second feature of allowing the dialogs to also be resize. Depending on who does the work for this, I am ok if the resizing portion is bumped down the road to a second feature.
- I think by default the dialog should reset its size/position each time it is shown. Though I think it would be nice to add in an opt-in feature that persists the last location.
- I think preventing the dialog from being dragged off screen (in the case of the default style) would be preferable. Beyond this, I would not restrict the movement of the dialog. In the case of the embedded dialog host, I would simply force the dialog to remain within the bounds of the parent container.
- I think once the user has dragged the dialog, it should hold that position and not move if the parent window moves. However if the user has not dragged the dialog then the existing re-positioning behavior should still work. I would be open to including some sort of re-link behavior where the dialog could maintain its current position and get re-positioned relative to the host window when it moves. However, I would consider this optional.
- Going along with that previous point, we should add in some programmatic means of resetting the dialogs position. In the case of the default style this should also re-enable the popup's re-positioning when the host window moves behavior.

Questions:
- How do we specify the region that the user should click and drag on to move the dialog? Should this simply be root `Card`? 
- For this we will need to consider that subsequent dialogs may vary in size. If we are showing from the last location do we show at the last X/Y (top left) position or based on the last centered position?

**Comment by AlexejLiebenthal** [comment]
> * There seems to be a subtle second feature of allowing the dialogs to also be resize. Depending on who does the work for this, I am ok if the resizing portion is bumped down the road to a second feature.

You're right this should be part of another issue. I doesn't thought about this, but this could also be nice.
To be clear: Everytime I mentioned resize/maximize/restore, I meant the window / dialog host.

> * I think by default the dialog should reset its size/position each time it is shown. Though I think it would be nice to add in an opt-in feature that persists the last location.

I fully agree on this.

> * I think preventing the dialog from being dragged off screen (in the case of the default style) would be preferable. Beyond this, I would not restrict the movement of the dialog. 

With screen, you mean the application window or the screen/monitor/desktop? 

> * I think once the user has dragged the dialog, it should hold that position and not move if the parent window moves. However if the user has not dragged the dialog then the existing re-positioning behavior should still work. I would be open to including some sort of re-link behavior where the dialog could maintain its current position and get re-positioned relative to the host window when it moves. However, I would consider this optional.

The repositioning behavior comes for free for the `PopupEx` style and as in the embedded style the grid is part of the visual tree, this is also I think already the default behavior than.
For holding the position a bit more work needs to be done. I would consider to have an additional `Thumb` on the DialogHost.Content, and then subtract the movements from the popup vertical / horizontal offset or the margins on the grid for the embedded one. Than there need to be also clarified what behavior the embedded control should have, if the dialog, would be go out of the viewport.
As mentioned before the dialog will be bounded to the edges, but when I move than the window back, is the dialog immediately moving or does it stay bounded till the window reaches the position, where the dialog was bounded the first time. Quite complicated to explain. I hope this helps:
[xdx] -> -[dxx] -> --[dxx] -> (-[xdx] -> [xxd]) or (-[dxx]  -> [xdx])
[ ] - window edges
x - window background
d - dialog
\- - window movement offset.

> * How do we specify the region that the user should click and drag on to move the dialog? Should this simply be root `Card`?

Hmm. I think, I would put the `Card` in a Grid and than put in the Grid a `Thumb` to get for a drag the drag offsets. Also I would apply a new `TemplatePart` for the `Thumb`. Maybe with some `DragAssist` than set the properties of the thumb. If not set the whole dialog should be clickable for drag. But I think the `DragAssist` should maybe done in a future/another issue.

> * For this we will need to consider that subsequent dialogs may vary in size. If we are showing from the last location do we show at the last X/Y (top left) position or based on the last centered position?

Good point. Maybe centered 🤔

**Comment by Keboo** [comment]
@AlexejLiebenthal sorry for the delay.

> With screen, you mean the application window or the screen/monitor/desktop?

I specifically mean the screen/monitor/desktop. Keeping in mind that a user may have more than one screen.

Regarding the offset. I "think" (-[xdx] -> [xxd]) makes the most sense but I might want to play with it to see.

**Comment by AlexejLiebenthal** [comment]
> @AlexejLiebenthal sorry for the delay.

No problem. I'm also quite busy these days/weeks. Hope I can find spare time to implement it soon.

> I specifically mean the screen/monitor/desktop. Keeping in mind that a user may have more than one screen.

I think this will be more difficult to implement, as `Popup` doesn't like to be split on screen edges.

> but I might want to play with it to see.

Same for me.
